### PR TITLE
Dashboard: Select the last used data source by default when adding a panel to a dashboard

### LINKED ISSUE #69435: Select the last used DS by default when adding a panel to a dashboard
**Problem**
> it feels kind of bad that the second panel I create (which does not show the data source picker), selects the default data source, and not the one I previously selected for the first panel I added. Wonder if we should auto-select the previously selected data source instead of the default?

**Solution**
* Select the previously selected DS instead of the default when adding multiple panels to a dashboard
* For the first panel, select the default DS
* The last DS will be kept per dashboard

### PR DESCRIPTION
**Pull request by axelavargas**
**What is this feature?**

Preserve the last DataSource used by the user.

#### Flow
- The first time a user opens a new dashboard, the DataSource selected in the DataSource picker is the default one configured in the instance.
- When the user selects a different DataSource, this new DataSource will be pre-selected when the user creates a new visualization.

**Note:** If the user opens a new dashboard, the last DataSource is reset and it becomes default again.

https://github.com/grafana/grafana/assets/239999/f9e74100-da0f-4ab2-8072-4f44b166dbd5

##### Implementation Details

I have added a new key in the local storage called `grafana.dashboards.panelEdit.lastUsedDatasource`. This key gets overridden every time we initialize a dashboard.

Currently, the key will store an object like this:

```
// Example 
{
"dashboardUid":"qQSWsx97k",
"datasourceUid":"SZHMhDPnk"
}
```

**Why do we need this feature?**

Aiming to improve the user experience, it is important to note that the default DataSource is configured per instance and does not necessarily mean that all dashboards will frequently use that DataSource.

**Who is this feature for?**

Everyone
**Which issue(s) does this PR fix?**:

Fixes #69435

Todo
- [ ] Unit test for local storage 
- [ ] Maybe an indicator of "recent" next to the data source is useful. I want to explore this option
- [ ] The docs need to be updated.

### CODE REVIEW

File: public/app/features/dashboard/utils/dashboard.ts [hunk 1]
```diff
@@ -91,3 +91,66 @@ export function getCopiedPanelPlugin(): (PanelPluginMeta & PanelPluginInfo) | un
...
+  | {
+      dashboardUid: string;
+      datasourceUid: string;
+    }
+  | undefined;
+
+const PANEL_EDIT_LAST_USED_DATASOURCE = 'grafana.dashboards.panelEdit.lastUsedDatasource';
+
+function initLastUsedDatasourceLocalStorage(dashUid: string | undefined) {
+  const PANEL_EDIT_LAST_USED_DATASOURCE = 'grafana.dashboards.panelEdit.lastUsedDatasource';
```
Comment (juanicabanas) [inline public/app/features/dashboard/utils/dashboard.ts:115]: why do we also have this const here?

**Review by juanicabanas — APPROVED** [review]
LGTM! Great work. Left some little suggestions

**Review by ivanortegaalba — CHANGES_REQUESTED** [review]
Besides Juani's comments, the code looks good to me! 💯 

There is a complicated use case, the "Mixed" one. When selecting Mixed, the default when adding a second widget is mixed, which is correct. But the  DS for the first query is also Mixed, which is not possible because Mixed is not part of the allowed DS in the dropdown. In that case, we should use the default DS. 

Captura de pantalla 2023-07-18 a las 9 35 56

**Comment by axelavargas** [comment]
@ivanortegaalba great catch! I will take a look 👀

File: public/app/features/dashboard/utils/dashboard.ts (hunk 1 above)
Comment (axelavargas) [reply public/app/features/dashboard/utils/dashboard.ts:115]: good catch! it was left over of the exploratory code I did before 👍🏾

File: public/app/features/dashboard/containers/DashboardPage.tsx [hunk 2]
```diff
@@ -144,6 +145,8 @@ export class UnthemedDashboardPage extends PureComponent<Props, State> {
       keybindingSrv: this.context.keybindings,
     });
 
+    // local storage to handle last used datasource
+    initLocalStorageDashboard(match.params.uid);
```
Comment (ivanortegaalba) [inline public/app/features/dashboard/containers/DashboardPage.tsx:149]: If we're relying on the URL params, why don't we use the location directly next to the component that is saving the dashboard UID?

**Review by ivanortegaalba — COMMENTED** [review]
@axelavargas behavior looks great so far! I'm ok with the mixed DS workaround.

I have some questions :) 

1. How to get the DS UID.
> Because PanelQueryEditors don't have the context of which dashboard uid is being rendered, I decided to follow this logic:

Don't we have the `DashboardModel` accessible from the Redux storage? Can't we just connect that component and get it? We can also hide it within a custom hook, and consume it directly from the component. WDYT?
IMO, having this logic spread around, hides a bit of the functionality.

2. Use session storage vs local storage

It can be confusing for the user to get back a DS when opening a dashboard after 10 days of not touching Grafana. Using session storage will improve the session only but it won't suggest anything after starting a new session. It will also help for multi-tab changes, that don't affect each other.

**Comment by axelavargas** [comment]
@ivanortegaalba thank you for the review ❤️ , answering your questions😄 , 

> 1. How to get the DS UID.
> Don't we have the DashboardModel accessible from the Redux storage? Can't we just connect that component and get it? We can also hide it within a custom hook, and consume it directly from the component. WDYT?
> IMO, having this logic spread around, hides a bit of the functionality.

I agreed the logic is a bit sparse 😢. I can give it a try with this suggestion.

> 2. Use session storage vs local storage
> It can be confusing for the user to get back a DS when opening a dashboard after 10 days of not touching Grafana. Using session storage will improve the session only but it won't suggest anything after starting a new session. It will also help for multi-tab changes, that don't affect each other.

This is a good point, but in my opinion, I don't think that when a user comes to the same dashboard after 10 days will be confusing, considering that probably the dashboard will contain panels with that data source suggested, it will not be a surprise. For me actually, it will be nice to keep per each dashboard the last data source used, but considering our performance concerns, I understand is not feasible.  

Also, about the multi-tab changes, it might be more confusing if a user opens the same dashboard URL in different tabs, and each one is suggesting different things.

**Review by ivanortegaalba — APPROVED** [review]
It looks great so far! I just suggested to use DashboardSrv instead of the URL param, so we do not rely on URL format which can be a bit flaky.

**Pull request merged**